### PR TITLE
fix(router): http headers sensitive for expressions and insensitive for traditional_compatible

### PR DESCRIPTION
**Pull request by chronolaw**
### Summary

KAG-2905

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

### Full changelog

* [Implement ...]

### Issue reference

Fix #_[issue number]_

### CODE REVIEW

File: kong/router/atc.lua [hunk 1]
```diff
@@ -570,6 +570,19 @@ do
   local tb_concat = table.concat
   local replace_dashes_lower = require("kong.tools.string").replace_dashes_lower
 
+  -- traditional_compatible is case insensitive
+  local tuning_header = string.lower
```
Comment (dndx) [inline kong/router/atc.lua:574]: The traditional compatible code already uses `lower(xxx)`, why is this still needed?

Comment (chronolaw) [reply kong/router/atc.lua:574]: It is used for calculating cache hash.

Comment (dndx) [reply kong/router/atc.lua:574]: Hmm, if we are already matching by lowercase, then does the cache hash matter? Shouldn't it also be lowercase?

Comment (chronolaw) [reply kong/router/atc.lua]: Removed `lower`.

Comment (chronolaw) [reply kong/router/atc.lua:574]: Add back `tuning_header()` and a test case to verify the insensitive cache calculation.

Comment (chronolaw) [reply kong/router/atc.lua]: remove `tuning_header()` and update the test case.

File: spec/01-unit/08-router_spec.lua [hunk 2]
```diff
@@ -5063,5 +5147,92 @@ do
...
+          route   = {
+            id = "e8fb37f1-102d-461e-9c51-6608a6bb8102",
+            expression = [[http.path == "/foo/bar" && lower(http.headers.test2) == "quote"]],
+            priority = 100,
+          },
+        },
+      }
+    end)
+
+    it("select() should match with case sensitive", function()
```
Comment (dndx) [inline spec/01-unit/08-router_spec.lua:5177]:
```suggestion
    it("select() should match with case sensitivity", function()
```

File: changelog/unreleased/kong/expression_http_headers_sensitive.yml [hunk 3]
```diff
@@ -0,0 +1,6 @@
+message: |
+    Enable case sensitive in `expressions` router flavor,
+    and make the `traditional_compatible` flavor more compatible
+    with the `traditional` flavor's behavior.
```
Comment (dndx) [inline changelog/unreleased/kong/expression_http_headers_sensitive.yml:4]:
```suggestion
    Header value matching in `expressions` router flavor are now case sensitive.
    This change does not have effect on `traditional_compatible` mode
    where header value match is always performed ignoring the case.
```

**Review by dndx — APPROVED** [review]

**Review by chobits — APPROVED** [review]

**Pull request merged**